### PR TITLE
str_duplicate: Don't share with a frozen shared string

### PR DESCRIPTION
**Pull request by XrXr**
This is a follow up for 3f9562015e651735bfc2fdd14e8f6963b673e22a.
Before this commit, it was possible to create a shared string which
shares with another shared string by passing a frozen shared string
to `str_duplicate`.

Such string looks like:

```
 --------                    -----------------
 | root | ------ owns -----> | root's buffer |
 --------                    -----------------
     ^                             ^   ^
 -----------                       |   |
 | shared1 | ------ references -----   |
 -----------                           |
     ^                                 |
 -----------                           |
 | shared2 | ------ references ---------
 -----------
```

This is bad news because `rb_fstring(shared2)` can make `shared1`
independent, which severs the reference from `shared1` to `root`:

```c
/* from fstr_update_callback() */
str = str_new_frozen(rb_cString, shared2);  /* can return shared1 */
if (STR_SHARED_P(str)) { /* shared1 is also a shared string */
    str_make_independent(str);  /* no frozen check */
}
```

If `shared1` was the only reference to `root`, then `root` can be
reclaimed by the GC, leaving `shared2` in a corrupted state:

```
 -----------                         --------------------
 | shared1 | -------- owns --------> | shared1's buffer |
 -----------                         --------------------
      ^
      |
 -----------                         -------------------------
 | shared2 | ------ references ----> | root's buffer (freed) |
 -----------                         -------------------------
```

Here is a reproduction script for the situation this commit fixes.

```ruby
a = ('a' * 24).strip.freeze.strip
-a
p a
4.times { GC.start }
p a
```

 - string.c (str_duplicate): always share with the root string when
   the original is a shared string.
 - test_rb_str_dup.rb: specifically test `rb_str_dup` to make
   sure it does not try to share with a shared string.

[Bug #15792]

### CODE REVIEW

File: string.c [hunk 1]
```diff
@@ -1493,34 +1493,29 @@ str_replace(VALUE str, VALUE str2)
...
     const VALUE flag_mask =
 	RSTRING_NOEMBED | RSTRING_EMBED_LEN_MASK |
 	ENC_CODERANGE_MASK | ENCODING_MASK |
 	FL_TAINT | FL_FREEZE
 	;
     VALUE flags = FL_TEST_RAW(str, flag_mask);
     VALUE dup = str_alloc(klass);
-    MEMCPY(RSTRING(dup)->as.ary, RSTRING(str)->as.ary,
-	   char, embed_size);
+    RSTRING(dup)->as = RSTRING(str)->as;
```
Comment (nobu) [inline string.c]: You may want to separate these assignments as another PR?

**Comment by nobu** [comment]
Good catch.

File: string.c (hunk 1 above)
Comment (nobu) [reply string.c]: Even another commit would be nice, of course.

Comment (XrXr) [reply string.c]: I dropped it since it's fairly unrelated :)

**Comment by XrXr** [comment]
@nobu could I get some more feedback?

File: test/-ext-/string/test_rb_str_dup.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+require 'test/unit'
+require '-test-/string'
+
+class Test_RbStrDup < Test::Unit::TestCase
+  def test_nested_shared_non_frozen
+    str = Bug::String.rb_str_dup(Bug::String.rb_str_dup("a" * 50))
+    assert Bug::String.shared_string?(str)
+    refute Bug::String.sharing_with_shared?(str), '[Bug #15792]'
```
Comment (nobu) [inline test/-ext-/string/test_rb_str_dup.rb]: Sorry to be late.
Could you use `assert_send [Bug::String, :shared_string?, str]` and `assert_not_send` for better failure messages?
